bloom-housing/bloom
# feat: new cron job for duplicates

This PR addresses #4208 


## Description

Creates a new and improved duplicates (application flagged set) cron job. This tackles the issues identified [here](https://docs.google.com/document/d/1h7zK3b466M2mfgj5vMjTaQKSL8STKUrH0t3MW-C1T4U/edit#heading=h.r0j0xbeefnoe). 

How the duplicates are grouped were decided on via a slack poll. Option 2 won https://exygy.slack.com/archives/C01Q4QG5R8Q/p1724708380283439. 

Since this will interfere with the listings that have had the old cron job run there is a new date that we set via env variable. Any listing that closes before that date/time will have the old cron job run against it whereas any listing closed after it will have the new one. This way any partner that has resolved applications but a new application comes through they will not have to start from the beginning again. And if the new close date is in the future (or not set) only the old job will run.

Questions:
- If a new application comes through that matches an existing flagged set do we need to set the flagged set status back to "pending" so that it needs to be re-resolved?

Notes:
- A new view is created that has all of the possible flagged sets in the system. A query is done against this to find the duplicates.
- We previously required the rule_key to be unique. The way we ensured that it was unique was that each key was prefixed with the listing id. This seemed unnecessary since we are already storing the listing id. But in order to make sure the rows are unique this PR also removes the unique constraint and adds a compound unique constraint to the listing_id and rule_key

## How Can This Be Tested/Reviewed?

This can be tested both purely with the backend as well as using the partner site.

Prerequisites for both ways:
- Reseed the backend (more applications have been added to the District View listing)
- Add new environment variables to .env for backend
  - DUPLICATES_CLOSE_DATE="2024-07-28 00:00 -08:00"
  - DUPLICATES_PROCESSING_CRON_STRING=0 * * * * 

Using the partner site:
- Sign into the partner site and go to the District View listing.
- Close the listing and go to the "applications" tab
- Notice the pending duplicate sets and resolve the applications as seen fit

Using the backend:
- After a reseed either wait for the cron job to run or go to the swagger doc and execute the cron job 
- Look in the DB and verify that the appropriate Application Flagged sets exist and along with the connected applications


## Author Checklist:


## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates
